File: framework/languages/dotnet/samples/ShoppingMallCheckout/Server/CommerceApi/Ports/Outbound/WorkflowPorts.cs
using ShoppingMallCheckout.Shared.Contracts;

namespace ShoppingMallCheckout.Server.CommerceApi.Ports.Outbound;

internal interface IOrderWorkflowRouter
{
    ValueTask<OrderState> StartAsync(
        StartOrderWorkflowReq command,
        CancellationToken cancellationToken);

    ValueTask<OrderState> ContinueAsync(
        ContinueOrderWorkflowReq command,
        CancellationToken cancellationToken);

    ValueTask<OrderState> RebuildProjectionAsync(
        RebuildOrderProjectionReq command,
        CancellationToken cancellationToken);
}

internal interface ICommerceApiPeerClient
{
    ValueTask<StartOrderRes> ForwardStartAsync(
        string ownerInstanceId,
        StartOrderReq request,
        CancellationToken cancellationToken);
}

internal sealed record CommerceApiInstanceOptions(string InstanceId);

File: framework/languages/node/samples/ShoppingMall.Ts/Server/CommerceApi/Ports/Outbound/workflow-ports.ts
import type {
  RebuildOrderProjectionReq,
  RebuildOrderProjectionRes,
  SeedPendingIdempotencyReq,
  StartOrderRes,
  StartOrderWorkflowReq,
  StartOrderWorkflowRes
} from '../../../../Shared/Contracts/messages';

interface OrderWorkflowCommandPort {
  startOrder(request: StartOrderWorkflowReq): Promise<StartOrderWorkflowRes>;

  rebuildProjection(request: RebuildOrderProjectionReq): Promise<RebuildOrderProjectionRes>;

  seedPendingIdempotency(request: SeedPendingIdempotencyReq): Promise<StartOrderRes>;
}

export type {
  OrderWorkflowCommandPort
};
